rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,392 @@
+- `match<Arrow, Arrow>`: produces sightings for the configured matchers
+
+Now consider the use case of constructing a matcher from CSV data:
+
+```bash
+vast exec 'from file:///tmp/blacklist.csv |
+           read csv |
+           build --extract=net.src.ip' > matcher.flatbuf
+```
+
+After having constructed a matcher, we can now use it in a pipeline for
+matching:
+
+```bash
+vast exec 'from kafka -t /zeek/conn |
+           read zeek |
+           match --state matcher.flatbuf --on=:addr
+```
+
+The idea is that `match` gets the opaque matcher state as input. (We may
+generalize this builder pattern to deploy arbitrarily complex detections, e.g.,
+pre-trained models.)
+
+#### Remote execution
+
+In the above example, we passed in a state file via `--state`. This is a
+different pattern from today's matchers that run remotely. Assuming that we have
+a mechanism for remote state management (specifically matcher state), we can
+apply our existing UX pattern of remote matcher management.
+
+Let's assume the matcher plugin provides the `matcher` command for state
+management, just like today. To load the matcher state into a remote matcher, we
+nothing would change:
+
+```bash
+vast matcher load foo < matcher.flatbuf
+```
+
+Assuming this operations updates the state of matcher `foo`, we can then
+reference it in a remote pipeline:
+
+```bash
+vast spawn 'live | match --on=:addr foo'
+```
+
+Here, `live` is a dummy operator that represents the ingest path.
+
+### Pipeline Management
+
+In many cases we want to run pipelines remotely, independent of where the client
+is. This begs the challenge: how do we manage pipelines? As with all other VAST
+functions, there could be a client command.
+
+```bash
+# Show all pipelines:
+vast pipeline list
+```
+
+Create remote pipelines:
+
+```bash
+vast pipeline create foo 'where #type == "suricata.flow"'
+vast pipeline list
+# Output:
+# foo: where #type == "suricata.flow"
+```
+
+The idea would be that only clients can manipulate server state. This fixes
+the mess of having server-side YAML that goes out of sync with the server
+state when new things happen, e.g., new schemas arrive or new matchers get
+spawned.
+
+From now on, everything is client-side managed *only*. No more server-side
+state configuration beyond options/settings. This is where you can
+have your YAML for declarative ops:
+
+```bash
+vast pipeline load < pipelines.yaml
+```
+
+If you want the current server-side state for reproducing pipelines elsewhere,
+just dump it (as YAML or JSON):
+
+```bash
+vast pipeline dump --json
+vast pipeline dump --yaml
+```
+
+The output is not to be confused with the settings in `vast.yaml`. The pipeline
+state is still persisted on the server, though, because it must survive
+restarts. Consequently, the state changes must be atomic and be applied
+in a WAL manner.
+
+### REST API
+
+In the future, it would also be nice to offer the same pipeline management
```

Comment:
I would like to better understand what "pipeline management" is in this context. I find it difficult to imagine a REST API that can express the composable nature of these operators properly (except for taking a single string representing the entire pipeline).

Reply:
I thought the parent section called `## Pipeline Management` makes this clear, no?

Reply:
Got it - it's the equivalent of managing stored procedures :-).

I assume the REST API would take a string representation of the the pipeline, or do you also envision a `json` object representation of the same thing? This would be somewhat similar to what I've seen with Mongo, and IIRC, also elastic.

Reply:
> Got it - it's the equivalent of managing stored procedures :-).

Exactly.

> or do you also envision a json object representation of the same thing?

That's a good idea. Eventually "yes", but I see it outside of this proposal.

A related thing: pipelines should of course convert into substrait query plans. :neckbeard:

Reply:
There's not much to do in this thread, resolving.